master/internal/command/events.go

```diff
@@ -44,17 +48,70 @@ type event struct {
 	LogEvent *string `json:"log_event"`
 }
 
+type logSubscribers = map[*actor.Ref]webAPI.LogsRequest // CHECK could we get different pointers to
+// the same ref? thus resulting in different map keys
+
+// GetEventCount is an actor message used to get the number of events in buffer.
+type GetEventCount struct{}
+
 type eventManager struct {
-	bufferSize int
-	buffer     *ring.Ring
-	closed     bool
-	seq        int
+	bufferSize   int
+	buffer       *ring.Ring
+	closed       bool
+	seq          int
+	isTerminated bool // DISCUSS If we don't want to don't keep track of this here we'd either need to
+	// expose event externally or define and send a new message. also the logActor might not receive
+	// the termination log message (based on LogRequest) if we rely on exposing Event..
+	logStreams logSubscribers
 }
 
 func newEventManager() *eventManager {
 	return &eventManager{
-		bufferSize: defaultEventBufferSize,
-		buffer:     ring.New(defaultEventBufferSize),
+		bufferSize:   defaultEventBufferSize,
+		buffer:       ring.New(defaultEventBufferSize),
+		logStreams:   make(logSubscribers),
+		isTerminated: false,
+	}
+}
+
+// QUESTION where do we house these utilities?
+func countNonNullRingValues(ring *ring.Ring) int {
+	// OPT we could use log_buffer here instead of a plain ring buffer.
+	count := 0
+	ring.Do(func(val interface{}) {
+		if val != nil {
+			count++
+		}
+	})
+	return count
+}
+
+// TODO inline this.
+func (e *eventManager) RemoveSusbscribers(ctx *actor.Context) {
+	for actor := range e.logStreams {
+		// OPT this will trigger a bunch of CloseStream message that'll come back to eventManager.
```

Comment:
That's fine.  This is a fundamental race condition when you have objects with unstructured lifetimes (structured = one starts before and ends after the other).  Either could try to close the other in either order, and they need to be able to tolerate that.

The [GObject memory model](https://developer.gnome.org/gobject/stable/gobject-memory.html) (part of glib, from the GNOME team) deals with this by explicitly separating object shutdown into two phases: "disposal" and "finalizing".  Disposal must tolerate being called many times, as two objects might try and trigger each other's disposal functions *from within* each other's disposal functions.  But "finalizing" is guaranteed to only happen once, after nobody holds a reference to you anymore.

In the actor system, this means when the actor is closed, all `Tell` and `Ask` operations will still complete, but nothing is actually added to the inbox.  That deals with tolerating the undefined shutdown order.  The finalization step is when golang GC's the actual actor inbox, which happens when nobody holds a reference to that actor anymore.

TL;DR: this comment isn't necessary haha

Reply:
thanks for the detailed explanation!